src/v/archival/segment_reupload.cc

```diff
@@ -0,0 +1,271 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "segment_reupload.h"
+
+#include "cloud_storage/partition_manifest.h"
+#include "logger.h"
+#include "storage/disk_log_impl.h"
+#include "storage/fs_utils.h"
+
+namespace archival {
+segment_collector::segment_collector(
+  model::offset begin_inclusive,
+  const cloud_storage::partition_manifest* manifest,
+  const storage::disk_log_impl* log,
+  size_t max_uploaded_segment_size)
+  : _begin_inclusive(begin_inclusive)
+  , _manifest(manifest)
+  , _log(log)
+  , _max_uploaded_segment_size(max_uploaded_segment_size) {}
+
+void segment_collector::collect_segments() {
+    if (_manifest->size() == 0) {
+        vlog(
+          archival_log.warn,
+          "No segments to collect for ntp {}",
+          _manifest->get_ntp());
+        return;
+    }
+
+    align_begin_offset_to_manifest();
+
+    if (_begin_inclusive >= _manifest->get_last_offset()) {
+        vlog(
+          archival_log.warn,
+          "Start offset {} is ahead of manifest last offset {} for ntp {}",
+          _begin_inclusive,
+          _manifest->get_last_offset(),
+          _manifest->get_ntp());
+        return;
+    }
+
+    do_collect();
+}
+
+segment_collector::segment_seq&& segment_collector::take_segments() {
```

Comment:
can the return value here be a value (not an r-value reference)?

also, if possible (and it isn't always possible / convenient), you can r-value ref qualify the method itself to require something like `std::move(collector).take_segments()`.

Reply:
it looks like all the cases of take_segments being called (i only seem it in tests) are right before the collector is destroyed, so this could be r-value-ref qualified.

Reply:
but if it isn't used anywhere except tests, you could just return a copy to keep things simple since perf isn't criticial.

Reply:
I changed it to return a copy.